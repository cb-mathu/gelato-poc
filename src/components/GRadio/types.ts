import { ExtractPropTypes } from 'vue';
import { VueComponentProps } from '../../types';
import { RADIO_PROPS } from './constants';

export type RadioProps = ExtractPropTypes<typeof RADIO_PROPS>;

export type MergedRadioProps = Partial<RadioProps & VueComponentProps>;
