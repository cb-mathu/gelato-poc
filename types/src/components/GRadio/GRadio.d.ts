declare const _default: import("vue").DefineComponent<{
    value: {
        type: (StringConstructor | NumberConstructor)[];
    };
    id: {
        type: StringConstructor;
        required: boolean;
    };
    class: {
        type: StringConstructor;
        required: boolean;
        default: string;
    };
    disabled: {
        type: BooleanConstructor;
    };
    autofocus: {
        type: BooleanConstructor;
        default: boolean;
    };
    autocomplete: {
        type: StringConstructor;
        default: string;
    };
    required: {
        type: BooleanConstructor;
        default: boolean;
    };
    readonly: {
        type: BooleanConstructor;
        default: boolean;
    };
    name: {
        type: StringConstructor;
        default: string;
    };
    modelValue: {
        type: (StringConstructor | BooleanConstructor | NumberConstructor)[];
        default: string;
    };
    placeholder: {
        type: StringConstructor;
        default: string;
    };
    modelModifiers: {
        default: () => {};
    };
    invalid: {
        type: BooleanConstructor;
    };
    errorMessage: {
        type: StringConstructor;
    };
}, () => JSX.Element, unknown, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, "update:modelValue"[], "update:modelValue", import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps, Readonly<{
    class: string;
    autofocus: boolean;
    disabled: boolean;
    name: string;
    placeholder: string;
    modelValue: string | number | boolean;
    autocomplete: string;
    required: boolean;
    readonly: boolean;
    modelModifiers: {};
    invalid: boolean;
} & {
    id?: string | undefined;
    value?: string | number | undefined;
    errorMessage?: string | undefined;
}>, {
    class: string;
    autofocus: boolean;
    disabled: boolean;
    name: string;
    placeholder: string;
    modelValue: string | number | boolean;
    autocomplete: string;
    required: boolean;
    readonly: boolean;
    modelModifiers: {};
    invalid: boolean;
}>;
export default _default;
