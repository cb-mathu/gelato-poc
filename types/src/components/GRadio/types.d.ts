import { ExtractPropTypes } from 'vue';
import { VueComponentProps } from '../../types';
import { RADIO_PROPS } from './constants';
export declare type RadioProps = ExtractPropTypes<typeof RADIO_PROPS>;
export declare type MergedRadioProps = Partial<RadioProps & VueComponentProps>;
