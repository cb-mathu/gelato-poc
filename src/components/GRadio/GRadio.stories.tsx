import { ref } from 'vue';
import GRadio from './GRadio';
import { MergedRadioProps } from './types';

export default {
  title: 'Component/Forms/GRadio',
  component: GRadio,
  argTypes: {
    name: {
      default: 'Name of the field',
    },
    label: {
      description: 'Label for the field',
      control: {
        type: 'text',
      },
    },
    disabled: {
      description: 'Disable the field',
      control: {
        type: 'boolean',
      },
    },
    autofocus: {
      description: 'Autofocus the field',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    required: {
      description: 'Field is required',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    readonly: {
      description: 'Field is readonly',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    modelValue: {
      description: 'v-model value',
      control: {
        type: 'text',
      },
    },
    value: {
      description: 'Default value',
      control: {
        type: 'text',
      },
    },
    id: {
      description: 'Id of the field',
      control: {
        type: 'text',
      },
    },
    class: {
      description: 'Class of the field',
      defaulValue: '',
      control: {
        type: 'text',
      },
    },
    invalid: {
      description: 'Field is invalid',
      defaulValue: false,
      control: {
        type: 'boolean',
      },
    },
    errorMessage: {
      description: 'Error message for invalid field (Applicable only if invalid is set)',
      defaulValue: '',
      control: {
        type: 'text',
      },
    },
  },
};

const Template = (args: any) => ({
  setup() {
    const consent1 = args.groupLabel1 || 'On';
    const consent2 = args.groupLabel2 || 'Off';
    let input = ref('On');

    const attributes = {
      disabled: args.disabled,
      readonly: args.readonly,
      invalid: args.invalid,
      errorMessage: args.errorMessage,
    };

    return () => (
      <div style="display:grid;grid-gap: 0.75rem;">
        <b>Selected: {input.value}</b>
        <div>
          <GRadio {...attributes} name="sep-radio-story" id={consent1} value={consent1} v-model={input.value}>
            {args.groupLabel1}
          </GRadio>
          <GRadio {...attributes} name="sep-radio-story" id={consent2} value={consent2} v-model={input.value}>
            {args.groupLabel2}
          </GRadio>
        </div>
      </div>
    );
  },
});

//Story
export const Story: any = Template.bind({});
Story.args = {
  groupLabel1: 'On',
  groupLabel2: 'Off',
};

// With Label
export const DefaultRadio = (props?: MergedRadioProps, slots?: any) => (
  <GRadio key="default" {...(props as any)}>
    {slots?.default?.() || 'On'}
  </GRadio>
);

// Disabled
export const DisabledRadio = () => (
  <GRadio key="disable" disabled>
    On
  </GRadio>
);

// Readonly
export const ReadonlyRadio = () => (
  <GRadio key="readonly" readonly>
    On
  </GRadio>
);

// Invalid
export const InvalidRadio = () => (
  <GRadio key="invalid" invalid errorMessage="This is an error message, ooh poor you!">
    On
  </GRadio>
);

// Radio group
export const RadioGroup = (props?: any) => {
  return {
    setup() {
      const consent1 = 'On';
      const consent2 = 'Off';
      let input = ref(props?.value || 'On');

      return () => (
        <div style="display:grid;grid-gap: 0.75rem;">
          <b>Selected: {input.value}</b>
          <div>
            <GRadio name="sep-radio-story" id={consent1} value={consent1} v-model={input.value}>
              {consent1}
            </GRadio>
            <GRadio name="sep-radio-story" id={consent2} value={consent2} v-model={input.value}>
              {consent2}
            </GRadio>
          </div>
        </div>
      );
    },
  };
};
