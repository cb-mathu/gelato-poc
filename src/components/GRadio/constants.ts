import { BASE_FORM_PROPS } from '../../utils/props';

export const RADIO_PROPS = {
  ...BASE_FORM_PROPS,
  value: {
    type: [Number, String],
  },
};
