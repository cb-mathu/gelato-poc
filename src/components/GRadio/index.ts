import GRadio from './GRadio';

export { GRadio };
