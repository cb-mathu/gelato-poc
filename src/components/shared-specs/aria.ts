/* istanbul ignore file */
import { render } from '@testing-library/vue';
import userEvent from '@testing-library/user-event';
import { screen } from '@testing-library/dom';

export const FormFieldAria = (Component: any, label: string, role: string) =>
  describe(`G${role.toUpperCase()}-Form Field ARIA`, () => {
    test('renders aria-labelledBy', async () => {
      render(Component());
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toHaveAttribute('aria-labelledBy');
    });

    test('renders aria-invalid', async () => {
      render(Component({ invalid: true }));
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toBeInvalid();
    });

    test('renders aria-errorMessage', async () => {
      const errorMessage = 'This is an error message, ooh poor you!';
      render(Component({ invalid: true, errorMessage: errorMessage }));
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toHaveErrorMessage(errorMessage);
    });
  });

export const ControlFieldAria = (Component: any, label: string, role: string) =>
  describe(`G${role.toUpperCase()}-Control Field ARIA`, () => {
    test('renders aria-checked', async () => {
      render(Component());
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByRole(role, { checked: true })).toBeTruthy();
    });
    FormFieldAria(Component, label, role);
  });

export const InputFieldAria = (Component: any, label: string, role: string) =>
  describe(`G${role.toUpperCase()}-Input Field ARIA`, () => {
    test('renders aria-describedBy', async () => {
      render(Component());
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toHaveAccessibleDescription();
    });
    FormFieldAria(Component, label, role);
  });
