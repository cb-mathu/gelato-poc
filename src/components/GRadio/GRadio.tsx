import { computed, defineComponent, InputHTMLAttributes } from 'vue';
import { useVModel, useId } from '../../composables';
import { RADIO_PROPS } from './constants';
import { GInputErrorMessage } from '../GInput';
import { useControlAria, useFormAria, useFormAriaIds } from '../../utils/accessibility';

export default defineComponent({
  name: 'GRadio',
  props: RADIO_PROPS,
  emits: ['update:modelValue'],

  setup(props, { attrs, slots }) {
    const internalValue = useVModel(props, 'modelValue');
    const { labelId, errorId } = useFormAriaIds(props.id);

    const propsWeControl = {
      id: props.id || useId('GRadio'),
    };

    return () => {
      const attributes = { ...attrs, ...props, ...propsWeControl };
      const checked = computed(() => props.modelValue === props.value).value;

      const ariaAttributes = {
        ...useFormAria({
          labelledBy: labelId,
          invalid: props.invalid,
          errorMessage: errorId,
        }),
        ...useControlAria({
          checked,
        }),
      } as InputHTMLAttributes;

      return (
        <div class="g-radio">
          <div class="g-radio__group">
            <input type="radio" {...attributes} {...ariaAttributes} v-model={internalValue.value} class="g-radio__element" />
          </div>
          <label class="g-radio__label" for={attributes.id} id={labelId}>
            {slots.default?.()}
          </label>
          {props.invalid && props.errorMessage && (
            <GInputErrorMessage class="g-radio__error" id={errorId} errorMessage={props.errorMessage} />
          )}
        </div>
      );
    };
  },
});
