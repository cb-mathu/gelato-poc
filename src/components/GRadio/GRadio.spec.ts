import { render } from '@testing-library/vue';
import { screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import * as stories from './GRadio.stories';
import { ControlFieldAria } from '../shared-specs/aria';

const { DefaultRadio, DisabledRadio, ReadonlyRadio, InvalidRadio, RadioGroup } = stories;

const testLabel = 'On';
const testSecondaryLabel = 'Off';
const testErroMessage = 'This is an error message, ooh poor you!';

describe('GRadio: ', () => {
  test('renders with default args', () => {
    const { container } = render(DefaultRadio());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders with label passed via slot', () => {
    render(DefaultRadio());
    expect(screen.getByText(testLabel)).toBeVisible();
  });

  test('renders in disabled state', () => {
    render(DisabledRadio());
    expect(screen.getByRole('radio')).toHaveAttribute('disabled');
  });

  test('renders in readonly state', () => {
    render(ReadonlyRadio());
    expect(screen.getByRole('radio')).toHaveAttribute('readonly');
  });

  test('renders in invalid state', () => {
    render(InvalidRadio());
    expect(screen.getByText(testErroMessage)).toBeVisible();
  });

  test('pressing tab will focus radio', () => {
    render(DefaultRadio());
    userEvent.tab();
    expect(screen.getByRole('radio')).toHaveFocus();
  });

  test('focuses radio when label is clicked', () => {
    render(DefaultRadio());
    const element = screen.getByText(testLabel);
    userEvent.click(element);
    expect(screen.getByRole('radio')).toHaveFocus();
  });

  test('does not update value in disabled state when label is clicked', () => {
    render(DefaultRadio({ disabled: true }));
    userEvent.click(screen.getByText(testLabel));
    expect(screen.getByRole('radio')).not.toBeChecked();
  });

  test('updates the state on click', async () => {
    render(DefaultRadio());
    await userEvent.click(screen.getByRole('radio'));
    expect(screen.getByRole('radio')).toBeChecked();
  });

  test('renders checked radio', async () => {
    render(DefaultRadio({ value: 'On', modelValue: 'On' }));
    expect(screen.getByLabelText(testLabel)).toBeChecked();
  });
});

describe('GRadio Group: ', () => {
  test('atmost one radio must be checked', async () => {
    render(RadioGroup());
    await userEvent.click(screen.getByLabelText(testLabel));
    await userEvent.click(screen.getByLabelText(testSecondaryLabel));
    expect(screen.getByLabelText(testSecondaryLabel)).toBeChecked();
    expect(screen.getByLabelText(testLabel)).not.toBeChecked();
  });

  test('renders default value for radio group', async () => {
    render(RadioGroup({ value: 'On' }));
    expect(screen.getByLabelText(testLabel)).toBeChecked();
  });
});

ControlFieldAria(DefaultRadio, testLabel, 'radio');
