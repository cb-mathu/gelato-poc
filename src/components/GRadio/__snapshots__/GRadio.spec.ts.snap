// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GRadio:  renders with default args 1`] = `
<div
  class="g-radio"
>
  <div
    class="g-radio__group"
  >
    <input
      aria-checked="false"
      aria-errormessage=""
      aria-invalid="false"
      aria-labelledby="label_1"
      autocomplete="off"
      class="g-radio__element"
      id="GRadio_4"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue=""
      name=""
      placeholder=""
      type="radio"
    />
  </div>
  <label
    class="g-radio__label"
    for="GRadio_4"
    id="label_1"
  >
    
    On
    
  </label>
  <!---->
</div>
`;
